### DISCUSSION TITLE
Query Regarding Support for sycl::fract Function in AdaptiveCpp

### DISCUSSION THREAD

**Discussion by blinkfrog** [AdaptiveCpp/AdaptiveCpp#1257] [Q&A] [answered]
Hello,

I am currently working on a project using AdaptiveCpp, and I have encountered an issue with the `sycl::fract` function. When I attempt to compile my code, I receive the following error:

```
C:/Temp/sycltest/SyclCmakeProject5Fract/blur.cpp:127:20: error: no matching function for call to 'fract'
127 | float value = sycl::fract(v5, &dummy);
| ^~~~~~~~~~~
```

This error suggests that the `sycl::fract` function might not be supported

My questions are:

Does AdaptiveCpp currently support the sycl::fract function?
If it is supported, is there a specific way I need to use it to avoid this compilation error? `float value = sycl::fract(v5);` doesn't work either.

In overall, this isn't a big problem as I can implement `fract` function by subtracting `floor` result from original value (and inverting the sign in the case of negative values), just wanted to ask if it is supported, as it is in the current SYCL standard.

Thank you in advance for your help!

**Comment by illuhad** [comment]
It's implemented. The second argument must be a `sycl::multi_ptr` object.

**Comment by blinkfrog** [comment]
Thank you. Could you please explain with some example?
According to https://www.khronos.org/files/sycl/sycl-2020-reference-guide.pdf, the signature of `fract`' is `Tf fract (Tf x, Tf *iptr)`. If I understand this correctly, then the type of the second argument should be the pointer of the same type as result and the first argument. In my case, it is:
```
float dummy;
float value = sycl::fract(v5, &dummy); //v5 is of float type
```
And this is how it works in oneAPI.
I've tried to use `sycl::multi_ptr`, but can't make it to work.

**Reply by illuhad** [reply, marked as answer]
I don't know where the reference guide pulls this signature from. It seems wrong to me. Here's what the specification actually says for fract:

> Constraints: Available only if Ptr is multi_ptr with ElementType equal to the same type as x and with Space equal to one of the writeable address spaces as defined above.

(See https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html and search for `fract` in your browser)

The following compiles:
```c++
#include <sycl/sycl.hpp>

int main() {
  sycl::queue q;
  q.single_task([=](){
    float f;
    sycl::fract(3.4f, sycl::multi_ptr<float, sycl::access::address_space::generic_space>{&f});
  }).wait();
}

```

However, upon checking again, I noticed that currently only the generic SSCP compiler implements this builtin (but implements it for all backends). While the compilation flows have infrastructure for handling it, they don't actually define it. So for all other compilation flows you will get a linker error.

**Comment by blinkfrog** [comment]
Thanks.
